File: src/FormulaInterpreter.Logic.VisualBasic/Calculator.vb
Public Module Calculator
    Function Solve(ByVal s As String) As Integer
        If s Is Nothing Then Throw New ArgumentNullException()
        Dim f = s.Trim()
        If f.Length < 3 Then Throw New InvalidOperationException()
        Dim sum As Integer = 0
        Dim op As Char = "+"c
        For i As Integer = 0 To f.Length - 1
            If (f(i) = "+"c OrElse f(i) = "-"c) AndAlso op = "n"c Then
                op = f(i)
                Continue For
            End If
            If Char.IsDigit(f(i)) AndAlso op <> "n"c Then
                If op = "+"c Then sum += CInt(Char.GetNumericValue(f(i)))
                If op = "-"c Then sum -= CInt(Char.GetNumericValue(f(i)))
                op = "n"c
                Continue For
            End If
            If f(i) = " "c OrElse (f(i) = "="c AndAlso i + 1 = f.Length) Then Continue For
            Throw New InvalidOperationException()
        Next
        If op <> "n"c Then Throw New InvalidOperationException()
        Return sum
    End Function
End Module

File: src/FormulaInterpreter.Logic.Csharp/Calculator.cs
using System;

namespace FormulaInterpreter.Logic.Csharp
{
    /// <summary>
    /// A calculator which is able to solve simple formulas.
    /// </summary>
    public static class Calculator
    {
        /// <summary>
        /// Solves a simple fomula.
        /// </summary>
        /// <param name="formula">Formula consisting of '+', '-', '=' and number tokens.</param>
        /// <returns>Result of the formula.</returns>
        public static int Solve(string formula)
        {
            var tokens = Lexer.Tokenize(formula);

            var tokenTree = Parser.BuildTokenTree(tokens);

            return tokenTree.GetValue();
        }

        public static int SolveBoring(string s)
        {
            if (s == null) throw new ArgumentNullException();
            var f = s.Trim();
            if (f.Length < 3) throw new InvalidOperationException();
            int sum = 0;
            char op = '+';
            for (int i = 0; i < f.Length; i++)
            {
                if ((f[i] == '+' || f[i] == '-') && op == 'n') { op = f[i]; continue; }
                if (char.IsDigit(f[i]) && op != 'n')
                {
                    if (op == '+') sum += (int)char.GetNumericValue(f[i]);
                    if (op == '-') sum -= (int)char.GetNumericValue(f[i]);
                    op = 'n';
                    continue;
                }
                if (f[i] == ' ' || (f[i] == '=' && i + 1 == f.Length)) continue;
                throw new InvalidOperationException();
            }
            if (op != 'n') throw new InvalidOperationException();
            return sum;
        }
    }
}
